vllm/model_executor/layers/quantization/modelopt.py

```diff
@@ -1125,29 +1311,77 @@ def apply(
                 expert_map=expert_map)
 
         if self.fused_experts is None:
-            # If no modular kernel is provided, use cutlass_moe_fp4 for TP case
-            # only (no EP).
-            from vllm.model_executor.layers.fused_moe.cutlass_moe import (
-                cutlass_moe_fp4)
-            out = cutlass_moe_fp4(
-                a=x,
-                w1_fp4=layer.w13_weight,
-                w2_fp4=layer.w2_weight,
-                w1_blockscale=layer.w13_blockscale_swizzled,
-                w2_blockscale=layer.w2_blockscale_swizzled,
-                g1_alphas=layer.g1_alphas,
-                g2_alphas=layer.g2_alphas,
-                a1_gscale=layer.w13_input_scale_quant,
-                a2_gscale=layer.w2_input_scale_quant,
-                topk_weights=topk_weights,
-                topk_ids=topk_ids,
-                m=x.shape[0],
-                n=layer.w2_weight.shape[2] * 2,
-                k=x.shape[1],
-                e=layer.w13_weight.shape[0],
-                device=x.device,
-                expert_map=expert_map,
-                apply_router_weight_on_input=apply_router_weight_on_input)
+            # If no modular kernel is provided, use cutlass_moe_fp4 or
+            # flashinfer low latency kernel for TP case only (no EP).
+            if self.flashinfer_moe_backend == "TensorRT-LLM":
+                import flashinfer
+                a1_gscale = layer.w13_input_scale_quant
+                hidden_states_fp4, hidden_states_scale_linear_fp4 = flashinfer.fp4_quantize(  # noqa: E501
+                    x,
+                    a1_gscale,
+                    is_sf_swizzled_layout=False,
+                )
+                from vllm.model_executor.models.llama4 import Llama4MoE
+                use_llama4_routing = custom_routing_function is Llama4MoE.custom_routing_function  # noqa: E501
+                routing_method_type = flashinfer.RoutingMethodType.DeepSeekV3
+                if use_llama4_routing:
+                    routing_method_type = flashinfer.RoutingMethodType.Llama4
```

Comment:
Is there any way we can assert or otherwise check that the model is deepseek? We have FP4 qwen and other moes that I don't want to silently use this